stl/src/atomic_wait.cpp

```diff
@@ -0,0 +1,285 @@
+    }
+#endif // _ATOMIC_WAIT_ON_ADDRESS_STATICALLY_AVAILABLE
+
+} // unnamed namespace
+
+_EXTERN_C
+bool __stdcall __std_atomic_wait_direct(const void* _Comparand, const size_t _Size, _Wait_context& _Context) noexcept {
+#if _ATOMIC_WAIT_ON_ADDRESS_STATICALLY_AVAILABLE == 0
+    if (_Acquire_wait_functions() < __std_atomic_api_level::__has_wait_on_address) {
+        return __std_atomic_wait_indirect(_Context);
+    }
+#endif // _ATOMIC_WAIT_ON_ADDRESS_STATICALLY_AVAILABLE == 0
+
+    if (!__crtWaitOnAddress(const_cast<volatile void*>(_Context._Storage), const_cast<void*>(_Comparand), _Size,
+            _Get_remaining_wait_milliseconds(_Context._Deadline))) {
+        _Assume_timeout();
+        return false;
+    }
+
+    return true;
+}
+
+void __stdcall __std_atomic_notify_one_direct(const void* const _Storage) noexcept {
+#if _ATOMIC_WAIT_ON_ADDRESS_STATICALLY_AVAILABLE == 0
+    if (_Acquire_wait_functions() < __std_atomic_api_level::__has_wait_on_address) {
+        __std_atomic_notify_one_indirect(_Storage);
+        return;
+    }
+#endif // _ATOMIC_WAIT_ON_ADDRESS_STATICALLY_AVAILABLE = 0
+
+    __crtWakeByAddressSingle(const_cast<void*>(_Storage));
+}
+
+void __stdcall __std_atomic_notify_all_direct(const void* const _Storage) noexcept {
+#if _ATOMIC_WAIT_ON_ADDRESS_STATICALLY_AVAILABLE == 0
+    if (_Acquire_wait_functions() < __std_atomic_api_level::__has_wait_on_address) {
+        __std_atomic_notify_all_indirect(_Storage);
+        return;
+    }
+#endif // _ATOMIC_WAIT_ON_ADDRESS_STATICALLY_AVAILABLE == 0
+
+    __crtWakeByAddressAll(const_cast<void*>(_Storage));
+}
+
+
+void __stdcall __std_atomic_unwait_direct(_Wait_context& _Context) noexcept {
+    return __std_atomic_unwait_indirect(_Context);
+}
+
+void __stdcall __std_atomic_notify_one_indirect(const void* const _Storage) noexcept {
+    auto& _Entry = _Atomic_wait_table_entry(_Storage);
+    AcquireSRWLockExclusive(&_Entry._Lock);
+    _Wait_context* _Context = _Entry._Wait_list_head._Next;
+    for (; _Context != &_Entry._Wait_list_head; _Context = _Context->_Next) {
+        if (_Context->_Storage == _Storage) {
+            WakeConditionVariable(&reinterpret_cast<CONDITION_VARIABLE&>(_Context->_Condition));
+            break;
+        }
+    }
+    ReleaseSRWLockExclusive(&_Entry._Lock);
+}
+
+void __stdcall __std_atomic_notify_all_indirect(const void* const _Storage) noexcept {
+    auto& _Entry = _Atomic_wait_table_entry(_Storage);
+    AcquireSRWLockExclusive(&_Entry._Lock);
+    _Wait_context* _Context = _Entry._Wait_list_head._Next;
+    for (; _Context != &_Entry._Wait_list_head; _Context = _Context->_Next) {
+        if (_Context->_Storage == _Storage) {
+            WakeAllConditionVariable(&reinterpret_cast<CONDITION_VARIABLE&>(_Context->_Condition));
+            break;
+        }
+    }
+    ReleaseSRWLockExclusive(&_Entry._Lock);
+}
+
+
+bool __stdcall __std_atomic_wait_indirect(_Wait_context& _Context) noexcept {

```

Comment:
How do you feel about doing the memcmp in here so that all of the "unwait"s can go away?

Reply:
Good idea.
This proposal indeed removes "unwait". Along with "unwait", the `bool` that remains from "phase" goes away.

I still don't see how list can be singly-linked. Note that we unblock list guard while waiting. (Also keep in mind that there may be both concurrent waits on the same atomic and hash-collided waits on different atomics, due to later, we cannot use separate lock and own the table entry exclusively)

A minor downside would be that `memcmp` will have runtime parameter, so it will not degenerate to register comparison or rolled out inline memcpy, it will be library call. Not a big deal though, comparing to RMWs for SRWLOCK we have here (I see at least four: one for acquire, one release, one for enter sleep, one for exit sleep).

Another minor downside is potential spurious exit due to padding bits (unless padding bits is fully reverted to the original design with RMW in `atomic_ref` constructor). It will be handled in header, but there would be unnecessary unwait. The downside is not present in _original design of padding bits_, it will also will not be present in another idea I currently have in mind.

----

This subject of "unwait" partly overlaps with `SRWLOCK`s of atomic in new ABI. 
Ideally, for non-lock-free atomics we share their inherent lock for both protecting data and waiting on `CONDITION_VARIABLE`, this optimizes a lot (so for `atomic` it will not even be from the  table). We can try havign this for `atomic_ref` when both `atomic_ref` and atomic wait changes are merged.

Reply:
With this change it is no longer needed to check Windows version in header. Yet I'd prefer to keep your improvement in preprocessor magic for potential usage in other places.

Reply:
Actually, can even skip `memcmp`: with wait list, all spurious wakes originate from condition variable itself, so I don't expect too many of them. Still will add it.

Reply:
Yeah, looks way better, and now you can tune returned `bool` for `__std_atomic_wait_direct` and apply `aliasobj`